hass_nabucasa/ice_servers.py

```diff
@@ -0,0 +1,143 @@
+"""Manage ICE servers."""
+
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Awaitable, Callable
+import logging
+import random
+import time
+from typing import TYPE_CHECKING
+
+from aiohttp import ClientResponseError
+from aiohttp.hdrs import AUTHORIZATION, USER_AGENT
+from webrtc_models import RTCIceServer
+
+if TYPE_CHECKING:
+    from . import Cloud, _ClientT
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class IceServers:
+    """Class to manage ICE servers."""
+
+    def __init__(self, cloud: Cloud[_ClientT]) -> None:
+        """Initialize ICE Servers."""
+        self.cloud = cloud
+        self._refresh_task: asyncio.Task | None = None
+        self._ice_servers: list[RTCIceServer] = []
+        self._ice_servers_listener: Callable[[], Awaitable[None]] | None = None
+        self._ice_servers_listener_unregister: Callable[[], None] | None = None
+
+    async def _async_fetch_ice_servers(self) -> None:
+        """Fetch ICE servers."""
+        if TYPE_CHECKING:
+            assert self.cloud.id_token is not None
+
+        async with self.cloud.websession.get(
+            f"https://{self.cloud.servicehandlers_server}/webrtc/ice_servers",
+            headers={
+                AUTHORIZATION: self.cloud.id_token,
+                USER_AGENT: self.cloud.client.client_name,
+            },
+        ) as resp:
+            resp.raise_for_status()
+
+            self._ice_servers = [
+                RTCIceServer(
+                    urls=item["urls"],
+                    username=item["username"],
+                    credential=item["credential"],
+                )
+                for item in await resp.json()
+            ]
+
+        if self._ice_servers_listener is not None:
+            await self._ice_servers_listener()
+
+    def _get_refresh_sleep_time(self) -> int:
+        """Get the sleep time for refreshing ICE servers."""
+        timestamps = [
+            int(server.username.split(":")[0])
+            for server in self._ice_servers
+            if server.username is not None and ":" in server.username
+        ]
+
+        if not timestamps:
+            return 3600  # 1 hour
```

Comment:
If there was no timestamps, should we even continue?

Reply:
This should only occur when the cloud returns some invalid data (successfully), and I think we should attempt fetching valid data at some point. That said, I changed this to return between 40 and 60 minutes to prevent a burst of requests at once from HA instances.

Reply:
If that is the use case I think we can set that to 12-24h.
It will still trigger on startup if faster is desired.